ocaml/typing/typecore.ml

```diff
@@ -6702,6 +6911,193 @@ and type_function
         params_contain_gadt = contains_gadt; newtypes = newtype :: newtypes;
         fun_alloc_mode; ret_info;
       }
+  | { pparam_desc = Pparam_val (arg_label, None, pat); pparam_loc } :: rest
+    when is_unpack (arg_label, pat) && could_be_functor env ty_expected ->
```

Comment:
If we remove this case then `splti_function_mty`  40 lines later would fail saying a functor was expected but it wasn't the case.
This way codes like 
```ocaml
let f : _ -> _ = (module M : S) = ...
```
Are not typed as module dependent functions but as a regular function taking a first-class module as argument